hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java

```diff
@@ -457,6 +461,37 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile, Hood
   private JavaRDD<WriteStatus> upsertRecordsInternal(JavaRDD<HoodieRecord<T>> preppedRecords, String commitTime,
       HoodieTable<T> hoodieTable, final boolean isUpsert) {
 
+    if (getConfig().getSchemaCheck()) {
+      // Ensure that the current writerSchema is compatible with the latest schema of this
+      // dataset.
+      // When inserting/updating data, we read records using the schema saved in the
+      // data/log files and convert them to the GenericRecords with writerSchema.
+      // Hence, we need to ensure that this conversion can take place without errors.
+      try {
+        SchemaUtil schemaUtil = new SchemaUtil(hoodieTable.getMetaClient());
+        MessageType savedParquetSchema = schemaUtil.getDataSchema();
+        Schema savedSchema = schemaUtil.convertParquetSchemaToAvro(savedParquetSchema);
+        Schema writerSchema = HoodieWriteHandle.createHoodieWriteSchema(config.getSchema());
+        if (! schemaUtil.isSchemaCompatible(savedSchema, writerSchema)) {
+          String msg = "WriterSchema is not compatible with the schema present in the Table";
+          LOG.error(msg);
```

Comment:
lets have this as a single ERROR statement?

Reply:
The same msg is also used to raise the exception a few lines below. So I did not duplicate the string and created a local variable.